src/main/java/com/farion/onlinebookstore/controller/CategoryController.java

```diff
@@ -0,0 +1,87 @@
+import org.springframework.data.domain.Pageable;
+import org.springframework.http.HttpStatus;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Category management", description = "Endpoints for managing categories")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/categories")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+    private final ConditionHolder conditionHolder;
+
+    @PostMapping
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @ResponseStatus(HttpStatus.CREATED)
+    @Operation(summary = "Create a new category", description = "Create a new category for books")
+    public CategoryDto save(@RequestBody @Valid CreateCategoryRequestDto requestDto) {
+        CategoryDto newCategory = categoryService.createCategory(requestDto);
+        conditionHolder.setCondition(ThreadLocal.withInitial(categoryService::getAllCategoryIds));
+        return newCategory;
+    }
+
+    @GetMapping("/{id}/books")
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN', 'ROLE_USER')")
+    @Operation(summary = "Get all books by category id",
+            description = "Get a list of books by specific category id")
+    public List<BookDtoWithoutCategoryIds> getBooksByCategory(@PathVariable Long id) {
+        return bookService.findAllByCategory(id);
+    }
+
+    @GetMapping("/{id}")
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN', 'ROLE_USER')")
+    @Operation(summary = "Get category by id", description = "Get category by specific id")
+    public CategoryDto findById(@PathVariable Long id) {
+        return categoryService.getById(id);
+    }
+
+    @GetMapping
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN', 'ROLE_USER')")
+    @Operation(summary = "Get all categories",
+            description = "Get a list of all available categories")
+    public List<CategoryDto> findAll(Pageable pageable) {
+        return categoryService.findAll(pageable);
+    }
+
+    @DeleteMapping("/{id}")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @Operation(summary = "Soft delete category by id",
+            description = "Soft delete category by specific id")
+    public void delete(@PathVariable Long id) {
+        categoryService.deleteById(id);
+        conditionHolder.setCondition(ThreadLocal.withInitial(categoryService::getAllCategoryIds));
```

Comment:
if it were to stay - `ConditionHolder` should have been part of a service, not a controller,
But let's just remove it, from here and other places as well